subsys/net/lib/wifi_provision/wifi_provision.c

```diff
@@ -0,0 +1,1058 @@
+	if (client < 0) {
+		LOG_ERR("Error in accept %d, try again", -errno);
+		return -errno;
+	}
+
+	request.accepted = client;
+
+	net_addr_ntop(client_addr.sin6_family, &client_addr.sin6_addr, addr_str, sizeof(addr_str));
+	LOG_DBG("[%d] Connection from %s accepted", request.accepted, addr_str);
+
+	http_parser_init(&request.parser, HTTP_REQUEST);
+
+	while (true) {
+		received = recv(request.accepted, buf + offset, sizeof(buf) - offset, 0);
+		if (received == 0) {
+			/* Connection closed */
+			ret = -ECONNRESET;
+			LOG_DBG("[%d] Connection closed by peer", request.accepted);
+			goto socket_close;
+		} else if (received < 0) {
+			/* Socket error */
+			ret = -errno;
+			LOG_ERR("[%d] Connection error %d", request.accepted, ret);
+			goto socket_close;
+		}
+
+		/* Parse the received data as HTTP request */
+		(void)http_parser_execute(&request.parser,
+					  &parser_settings, buf + offset, received);
+
+		total_received += received;
+		offset += received;
+
+		if (offset >= sizeof(buf)) {
+			offset = 0;
+		}
+
+		/* If the HTTP request has been completely received, stop receiving data and
+		 * proceed to process the request.
+		 */
+		if (request.received_all) {
+			handle_http_request(&request);
+			break;
+		}
+	};
+
+socket_close:
+	LOG_DBG("Closing listening socket: %d", request.accepted);
+	(void)close(request.accepted);
+
+	return ret;
+}
+
+static int setup_server(int *sock, struct sockaddr *bind_addr, socklen_t bind_addrlen)
+{
+	int ret;
+	int enable = 1;
+
+	*sock = socket(bind_addr->sa_family, SOCK_STREAM, IPPROTO_TLS_1_2);
+	if (*sock < 0) {
+		LOG_ERR("Failed to create socket: %d", -errno);
+		return -errno;
+	}
+
+	sec_tag_t sec_tag_list[] = {
+		CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG,
+	};
+
+	ret = setsockopt(*sock, SOL_TLS, TLS_SEC_TAG_LIST, sec_tag_list, sizeof(sec_tag_list));
+	if (ret < 0) {
+		LOG_ERR("Failed to set security tag list %d", -errno);
+		return -errno;
+	}
+
+	ret = setsockopt(*sock, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(enable));
+	if (ret) {
+		LOG_ERR("Failed to set SO_REUSEADDR %d", -errno);
+		return -errno;
+	}
+
+	ret = bind(*sock, bind_addr, bind_addrlen);
+	if (ret < 0) {
+		LOG_ERR("Failed to bind socket %d", -errno);
+		return -errno;
+	}
+
+	ret = listen(*sock, 1);
+	if (ret < 0) {
+		LOG_ERR("Failed to listen on socket %d", -errno);
+		return -errno;
+	}
+
+	return ret;
+}
+
+/* Processing incoming IPv4 clients */
+static int process_tcp4(void)
+{
+	int ret;
+	struct sockaddr_in addr4 = {
+		.sin_family = AF_INET,
+		.sin_port = htons(CONFIG_WIFI_PROVISION_TCP_PORT),
+	};
+
+	ret = setup_server(&request.socket, (struct sockaddr *)&addr4, sizeof(addr4));
+	if (ret < 0) {
+		LOG_ERR("Failed to create IPv4 socket %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("Waiting for IPv4 HTTP connections on port %d", CONFIG_WIFI_PROVISION_TCP_PORT);
+
+	while (true) {
+
+		/* Process incoming IPv4 clients */
+		ret = process_tcp(AF_INET);
```

Comment:
I believe we should handle the ECONNRESET as a special case and continue staying in the loop. I managed to crash the provisioning sample by sending a bad request from httpie desktop client. It was because of the ECONNRESET returned by `process_tcp()` function.